src/core/qgsvectorlayerutils.cpp

```diff
@@ -734,13 +734,22 @@ void QgsVectorLayerUtils::matchAttributesToFields( QgsFeature &feature, const Qg
   feature.setFields( fields );
 }
 
-QgsFeatureList QgsVectorLayerUtils::makeFeatureCompatible( const QgsFeature &feature, const QgsVectorLayer *layer )
+QgsFeatureList QgsVectorLayerUtils::makeFeatureCompatible( const QgsFeature &feature, const QgsVectorLayer *layer, QgsFeatureSink::SinkFlags sinkFlags )
 {
   QgsWkbTypes::Type inputWkbType( layer->wkbType( ) );
   QgsFeatureList resultFeatures;
   QgsFeature newF( feature );
   // Fix attributes
   QgsVectorLayerUtils::matchAttributesToFields( newF, layer->fields( ) );
+
+  if ( sinkFlags & QgsFeatureSink::RegeneratePrimaryKey && layer->storageType() == QLatin1String( "GPKG" ) )
+  {
+    // drop incoming fid value, let it be regenerated
+    const int fidIndex = layer->fields().lookupField( QStringLiteral( "fid" ) );
```

Comment:
you should probably use the name of the first field returned by the provider instead of the hard coded "fid", in case it would have been named differently than the default.

Question (related to my last email): why do we need that hack for GPKG, but not with the Postgres provider when it exposes a integer primary key column ?

Reply:
Hmm - I'll need to dig deeper. My first guess at the difference proved wrong...

Reply:
Ok, fixed. I've made this method compatible with non-ogr/non-gpkg providers too.